tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+			idx += n + int(dataLen)
+		}
+	}
+
+	return nil
+}
+func (w *HeadReadWriter) Close() error {
+	w.bsMtx.Lock()
+	defer w.bsMtx.Unlock()
+
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	if err := w.dirFile.Close(); err != nil {
+		return err
+	}
+
+	return closeAllFromMap(w.cs)
+}
+
+// WriteChunk writes chunk in the following format:
+// | Series Ref <8B> | MinT <8B> | MaxT <8B> | Chunk Encoding <1B> | Chunk Data Length <varint> | Chunk Data |
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (w *HeadReadWriter) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chunkRef uint64, err error) {
+	w.wbufLock.Lock()
+	defer w.wbufLock.Unlock()
+
+	if w.shouldCutSegment(len(chk.Bytes())) {
+		if err := w.cut(); err != nil {
+			return 0, err
+		}
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], seriesRef)
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(mint))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(maxt))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the segment index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the segment index and
+	// the lower 4 bytes are for the segment offset where to start reading this chunk.
+	chunkRef = w.chunkRef(uint64(w.seq()), uint64(w.n))
+
+	w.buf[0] = byte(chk.Encoding())
+	if err := w.write(w.buf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(w.buf[:], uint64(len(chk.Bytes())))
+	if err := w.write(w.buf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := w.write(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	return chunkRef, w.wbuf.Flush()
+}
+
+func (w *HeadReadWriter) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int) bool {
+	return w.n == 0 || // First segment
+		// TODO: tune this boolean, cutting a segment for only 1 chunk would be inefficient.
+		(time.Since(w.curFileStartTime) > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
+}
+
+func (w *HeadReadWriter) seq() int {
+	return w.curFileSequence
+}
+
+func (w *HeadReadWriter) write(b []byte) error {
+	n, err := w.wbuf.Write(b)
+	w.n += int64(n)
+	return err
+}
+
+func (w *HeadReadWriter) cut() (err error) {
+	// Sync current tail to disk and close.
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, headChunksFormatV1, false, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	w.n = int64(n)
+	oldSeq := w.curFileSequence
+	w.curFileSequence = seq
+
+	// Write current time in milliseconds for mint and 0 for maxt.
```

Comment:
Why are we writing a walltime inside the tsdb?

Reply:
Whoops, had not thought much when I first implemented, it's definitely wrong. I will combine the suggestion of taking maxt from samples and write mint from samples too by changing it to `cut(t)`.